### PR TITLE
Add methodology documentation

### PR DESCRIPTION
**Pull request by spahrenk**
## Description

- The ppt presentation was transcribed to an md with minor changes .
- A null_implementation_test was added.

## Author Progress Checklist:

- [x] Open draft pull request
    - [x] Make title clearly understandable in a standalone change log context
    - [x] Assign yourself the issue
    - [x] Add at least one label (enhancement, bug, or maintenance)
    - [x] Link the issue(s) addressed by this PR (under "Development" in the sidebar menu) 
- [x] Make code changes (if you haven't already)
- [x] Self-review of code
    - [x] My code follows the style guidelines of this project
    - [x] I have added comments to my code, particularly in hard-to-understand areas
    - [x] I have only committed the necessary changes for this fix or feature
    - [x] I have made corresponding changes to the documentation
    - [x] My changes generate no new warnings
    - [x] I have ensured that my fix is effective or that my feature works as intended by:
        - [x] exercising the code changes in the test framework, and/or
        - [x] manually verifying the changes (as explained in the the pull request description above)
    - [x] My changes pass all local tests
    - [x] My changes successfully passes CI checks
    - [x] Add any unit test for proof and documentation.
    - [ ] Merge in main branch and address resulting conflicts and/or test failures.
- [ ] Move pull request out of draft mode and assign reviewers
- [ ] Iterate with reviewers until all changes are approved
    - [ ] Make changes in response to reviewer comments
    - [ ] Merge in main branch and address resulting conflicts and/or test failures.
    - [ ] Re-request review in GitHub

## Reviewer Checklist:

 - [ ] Read the pull request description
 - [ ] Perform a code review on GitHub
 - [ ] Confirm all CI checks pass and there are no build warnings
 - [ ] Pull, build, and run automated tests locally
 - [ ] Perform manual tests of the fix or feature locally
 - [ ] Add any review comments, if applicable
 - [ ] Submit review in GitHub as either
     - [ ] Request changes, or
     - [ ] Approve
 - [ ] Iterate with author until all changes are approved

### CODE REVIEW

File: docs/discussion.md [hunk 1]
```diff
@@ -0,0 +1,197 @@
+# **Interpolation**
+
+## **How the btwxt library works**
+## Big Ladder Software
+
+## 1-D Linear
+We do not know the true function $f$. We only know values of $f$ at a set of grid points $x_i$.
+
+To find $f(x)$ for $x$ in $(x_0,x_0)$,
```
Comment (tanaya-mankad) [inline docs/discussion.md:9]: Did you mean (x_0, x_1)?

File: docs/figs/fig_1Dcubic.png
Comment (tanaya-mankad) [inline docs/figs/fig_1Dcubic.png:1]: These figures are super. We also need to check in the source code used to generate them.

File: docs/discussion.md [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+\end{align*}$$
+
+
+<p align="center">
+    <img width="70%" src="./figs/fig_2Dlinear.png"> 
+</p>
+
+Four independent terms, one for each vertex.
+
+## 2-D Linear pseudocode
```
Comment (tanaya-mankad) [inline docs/discussion.md:42]: I think markdown supports a "code block" formatting, which we could use for pseudocode. - indenting the lines, maybe.

File: docs/discussion.md [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+Define:   
+     $c_{0}=2\mu^3-3\mu^2+1, \quad  c_{1}=-2\mu^3+3\mu^2$   
+     $d_{0}=\mu^3-2\mu^2+\mu, \quad d_{1}=\mu^3-\mu^2$  
+     $s_0=\delta_x/(x_1-x_{-1}), \quad s_1=\delta_x/(x_2-x_0)$
+
+Rewrite the polynomial:   
+ $f(\mu)=c_{0}\cdot f(0)+c_{1}\cdot f(1)+d_{0}\cdot s_0\cdot (f(1)-f(-1))+d_{1}\cdot s_1\cdot (f(2)-f(0))$  
+
+Group the coefficients:
```
Comment (tanaya-mankad) [inline docs/discussion.md:133]: :)

File: docs/discussion.md [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        case 2: weight *= d[1] $`*`$ s[1];  
+    f[x] += f[i] $\*\$ weight;  
+return f[x];
+
+Terms $c_{0}$ and $c_{1}$ are referred to programmatically as *interpolation coefficients*.
+
+The terms $d_{0}\cdot s_0$ and  $d_{1}\cdot s_1$ are referred to as *cubic_slope_coefficients*.
+
+## 2-D Cubic, with pictures
+There are 4<sup>N</sup> points to weigh. Number of calculations grows faster than that.
```
Comment (tanaya-mankad) [inline docs/discussion.md:154]: A few sentences seem to have gotten fragmented; these just need to read more fluidly.

File: docs/figs/btwxt presentation.pptx
Comment (tanaya-mankad) [inline docs/figs/btwxt presentation.pptx:1]: I think I made a mistake in viewing the figures - I thought the png files were from your own code that had generated  plots of the different cubic splines. If so, then you'd check in that code (Python or whatever). The pptx file is binary and I don't think it makes sense to commit it to the repository. @nealkruis what's your opinion?

File: docs/discussion.md [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+If 2-D, is is a rectangle
+
+If 3-D, it is a rectangular prism.
+
+The number of vertices of the hypercube = 2<sup>N</sup>.
+
+## N-D Linear pseudocode
+```c++
+f(x) = 0;  
+for each vertex x[i][j] in the hypercube:
```
Comment (tanaya-mankad) [inline docs/discussion.md:68]: This pseudocode is maybe useful, but it's not descriptive enough. What does `dim` range over? `vertex_weight` could use an earlier mention. `i` and `j` don't imply N dimensions to me, which is I think what we're after here (the slides used `...k` as well, but then didn't carry it through the loop).

File: docs/discussion.md [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+\end{align*}$$
+
+
+<p align="center">
+    <img width="70%" src="./figs/fig_2Dlinear.png"> 
+</p>
+
+Four independent terms, one for each vertex.
+
+## 2-D Linear pseudocode
```
Comment (tanaya-mankad) [inline docs/discussion.md:42]: Actually I think the 2-D linear pseudocode is both distracting and anyway you'd unroll the loop if you wrote it as real code, not think of it in this conditional format. Maybe we remove this?

File: docs/discussion.md [hunk 7]
```diff
@@ -0,0 +1,205 @@
+# **Interpolation**
+
+## **How the btwxt library works**
+## Big Ladder Software
+
+## 1-D Linear
```
Comment (tanaya-mankad) [inline docs/discussion.md:6]: Should there be a hyphen between the "1" and "D"? It looks funny but I guess we'd want it for "N-D"...

**Review by tanaya-mankad — CHANGES_REQUESTED** [review]

File: docs/discussion.md [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+    <img width="60%" src="./figs/fig_curve.png"> 
+</p>
+
+There are several approaches to interpolate a discrete set of known function values. For any single axis, **btwxt** supports both *linear* and *cubic* interpolation. The mathematical bases for these methods are described in this document. 
+
+**btwxt** also enables the interpolation of data sets that span multiple dimensions in parameter space. This requires combining the variation of the function with respect to each axis. This document also contains a description of the algorithm used to perform this N-dimensional interpolation.
+    
+##  Linear Interpolation
+Linear interpolation along an axis is often sufficient if either of the following conditions apply:
+1) the known funciton values are closely spaced, such that variations between neighboring points are within acceptable tolerances, or
```
Comment (tanaya-mankad) [inline docs/discussion.md:16]: typo: funciton

File: docs/discussion.md [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+```math
+\begin{align*}
+f(\mu)=&-d_{0}\cdot s_0&\cdot\quad &f(-1)\\
+&+(c_{0} -d_{1}\cdot s_1)&\cdot \quad &f(0)\\
+&+(c_{1}+d_{0}\cdot s_0)&\cdot\quad &f(1)\\
+&+d_{1}\cdot s_1&\cdot\quad &f(2)
+\end{align*}
+```
+
+Terms $c_{0}$ and $c_{1}$ are referred to as *interpolation coefficients*.
```
Comment (tanaya-mankad) [inline docs/discussion.md:170]: interpolation_coefficients (with the underscore, as below)

**Review by tanaya-mankad — CHANGES_REQUESTED** [review]
Really minor change requests this time. I really like how this flows; I can understand our assumptions and the algebra has just the right amount of explicitness for my personal taste!

**Review by spahrenk — COMMENTED** [review]
done

**Review by tanaya-mankad — APPROVED** [review]

File: docs/discussion.md [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+
+Terms $c_{0}$ and $c_{1}$ are referred to as *interpolation_coefficients*.
+
+The terms $d_{0}\cdot s_0$ and  $d_{1}\cdot s_1$ are referred to as *cubic_slope_coefficients*.
+
+<p align="center">
+    <img width="70%" src="./figs/fig_1D_cubic.png"> 
+</p>
+
+The above formulation allows cubic interpolation using the four function values associated with the grid-point coordinates in the vicinity of the queried coordinate. Notice that our slope estimations must be revised for edge cases, i.e., when either $x_0$ or $x_1$ represent the smallest or largest values, respectively, of the provided grid-axis coordinates. These cases can be handled by assigning the slope $(f(1)-f(0))/(x_1-x_0)$ to either $f'(0)$ or $f'(1)$, if either $x_{-1}$ or $x_2$, respectively, are outside the valid interpolation range.
```
Comment (nealkruis) [inline docs/discussion.md:178]: I wonder if we should break the discussion about edges into a new subheading?

File: docs/discussion.md [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+
+&ensp; &ensp; $s_0\gets 1$, $f(-1)\gets f(0)$
+
+are applied. If $x_1$ is the maximum coordinate, the assignments
+
+&ensp; &ensp; $s_1\gets 1$ , $f(2)\gets f(1)$
+
+are applied. 
+
+**btwxt** also enables *extrapolation* for estimation of the function outside the valid interpolation range. Two methods for extrapolation are available: *constant* and *linear*. If constant is selected, the function value for extrapolation is assumed to equal that of the nearest endpoint. A linear extrapolation returns the function value of the nearest endpoint at the grid-point coordinate for that endpoint, and has the slope assigned to that endpoint.
```
Comment (nealkruis) [inline docs/discussion.md:190]: Separate heading for extrapolation? This isn't specific to cubic interpolation. Maybe this should be moved after we're done discussing cubic interpolation?

File: docs/discussion.md [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+
+Cubic interpolation along both axes of a 2-D system involves $4^2=16$ points, as shown below.
+
+<p align="center">
+    <img width="50%" src="./figs/fig_2D_cubic_pic1.png"> 
+</p>
+
+As for the 2-D linear case, the coefficient for each vertex is the product of the coefficients computed for each axis, separately. Therefore, we first find the $2\cdot4=8$ single-axis coefficients, then we evaluate each of the $4^2=16$ pair products, which are the combined weighting factors for the 2-D vertices.
```
Comment (nealkruis) [inline docs/discussion.md:198]: Should this say "As for the 2-D cubic case..."? What coefficients are we referring to here? I think we're talking about weighting factors (per the following paragraph)?

File: docs/discussion.md [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+
+For example, let us label the weighting factors for axis-0 as u<sub>i</sub>. Those for axis-1 we will label as  v<sub>j</sub>. Then the combined weighting factor for vetex i,j is given by w<sub>i,j</sub>= u<sub>i</sub> $\cdot$ v<sub>j</sub>
+
+<p align="center">
+    <img width="50%" src="./figs/fig_2D_cubic_pic2.png"> 
+</p>
+
+Again, the extension to the N-D case is straightforward. If cubic interpolation is applied to all axes, a hypercube about each queried coordinate will contain 4<sup>N</sup> vertices.
+There are four coefficients along each axis, and 4N total. Each vertex is weighted by the product among all axes of the weighting factors for each axis.
```
Comment (nealkruis) [inline docs/discussion.md:207]: Missing superscript?

File: docs/discussion.md [hunk 14]
```diff
@@ -0,0 +1,232 @@
+# btwxt 
+## *Big Ladder Software*
```
Comment (nealkruis) [inline docs/discussion.md:2]: Maybe replace these lines with a single title that better describes the content of this document? The file name is "discussion.md", but I feel that isn't the most descriptive name. What if we call the file "methodology.md", and make the title "btwxt Methodology"?

**Review by nealkruis — CHANGES_REQUESTED** [review]
This is very close! Thanks, @spahrenk!

Mostly minor details here.

Do we want to include source code changes here too?

File: docs/discussion.md [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+
+&ensp; &ensp; $f'(0)=(f(1)-f(-1))/(x_1-x_{-1})$
+
+&ensp; &ensp; $f'(1)=(f(2)-f(0))/(x_2-x_0)$
+
+<p align="center">
+    <img width="70%" src="./figs/fig_slope_est.png"> 
+</p>
+
+We define:
```
[the remark this answers is not in the document]
Comment (nealkruis) [reply docs/discussion.md]: The corresponding addition doesn't appear to be rendering correctly.

File: docs/discussion.md (hunk 12 above)
Comment (nealkruis) [reply docs/discussion.md]: This doesn't seem to be clarified in the changes. I suspect "coefficient" is really talking about the "weighting factor".

File: docs/discussion.md (hunk 13 above)
Comment (nealkruis) [reply docs/discussion.md]: Still shows `4N`, when the "N" should be superscript.

File: docs/methodology.md [hunk 16]
```diff
@@ -0,0 +1,294 @@
...
+$$
+
+For cubic interpolation, the `weighting factors` are
+
+&ensp; &ensp; $W_{-1}=-D_0, \quad  W_0=C_0-D_1, \quad W_1=C_1+D_0, \quad W_2=D_1$
+
+
+<p align="center">
+    <img width="60%" src="./figs/fig_cubic.png"> 
+</p>
```
Comment (nealkruis) [inline docs/methodology.md:222]: This still doesn't look like it's matching the approximated slopes from the previous figure.

File: docs/methodology.md [hunk 17]
```diff
@@ -0,0 +1,294 @@
...
+\begin{align*}
+f(\mu_0,\mu_1)\quad=
+\quad&(1-\mu_0)\cdot (1-\mu_1)&\cdot \quad f(0,0)\\ 
++&\quad\mu_0\cdot (1-\mu_1)&\cdot \quad f(1,0)\\
++&\quad(1-\mu_0)\cdot \mu_1&\cdot  \quad f(0,1)\\
++&\quad\mu_0 \cdot \mu_1&\cdot  \quad f(1,1)
+\end{align*}
+$$
+
+The coefficient multiplying each function value comprises the *weight* of the *vertex* located at the corresponding 2-D grid-point coordinates. The weight for that vertex is simply the product of the `weighting factors` associated with each of the two axes. Notice that there are two unique `weighting factors` (i.e., $\mu_i$ and $1-\mu_i$) associated with each axis, so we must evaluate $2+2=4$ single-axis weights, in total. These are then combined to form $2\cdot 2=4$ weights for the full rectangle.
```
Comment (tanaya-mankad) [inline docs/methodology.md:84]: 👍

File: docs/figs/fig_2D_linear.png
Comment (tanaya-mankad) [inline docs/figs/fig_2D_linear.png:1]: I kinda liked the first image, but probably the new one is more descriptive. :)

**Review by tanaya-mankad — APPROVED** [review]

**Review by tanaya-mankad — COMMENTED** [review]
I don't have any context for what this commit was intended to accomplish, so I can't give a relevant review. In the future, you could add an explanatory comment of your own to the "conversation" section of the PR, and/or be a little more descriptive in the commit message (Also, please use the internal message style starting with a capital letter: “{Present-tense action} {on the thing this commit fixes}. “

**Comment by spahrenk** [comment]
Curve colors had been reversed in the most recent commit. An attempt was made to correct this.

**Comment by spahrenk** [comment]
Replaced actual slopes with estimated slope son cubic figure.

**Review by nealkruis — APPROVED** [review]

**Pull request merged**